### PR TITLE
Improve logging mechanism

### PR DESCRIPTION
**Pull request by tsorya**
adding network manager log from all nodes to network manager log file
changing writePodsLogs behaviour -> write to log file on success , write to log and stdout on failure
adding print node connections, devices and ip status before and after each test

Signed-off-by: Igal Tsoiref <itsoiref@redhat.com>

**What this PR does / why we need it**:
1. Adding option to write to log files.
2. Adding Network manager logs that are printed to <test>_netmanager.log
3. Pod logs will be added to <test>_pod.log and will be printed to stdout only on test failure
4. adding print node connections, devices and ip status before and after each test ()
**Special notes for your reviewer**:
Maybe commit is too big? i can try to split it

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -81,4 +81,8 @@ tags
 ### VisualStudioCode ###
 .vscode/*
 .history
+
+### Test logs ###
```
Comment (dankenigsberg) [inline .gitignore]: I have a personal vendeta against comments that add zero information.

Comment (tsorya) [reply .gitignore]: thought it is a convention here

Comment (dankenigsberg) [reply .gitignore]: maybe it's just me. stating the obvious is tiring.

File: Makefile [hunk 2]
```diff
@@ -101,6 +101,7 @@ test/unit: $(GINKGO)
 test/e2e: $(OPERATOR_SDK)
 	# We have to unset mod=vendor here since operator-sdk is already
 	# building with it, and go tool fail if it's specified twice
+	mkdir -p test_logs/e2e
```
Comment (phoracek) [inline Makefile]: On CI, we should copy this folder to exported artifacts, so it is available in the prow build. See https://github.com/nmstate/kubernetes-nmstate/blob/master/automation/check-patch.setup.sh#L12

File: automation/check-patch.setup.sh [hunk 3]
```diff
@@ -12,5 +12,6 @@ export TMP_PROJECT_PATH=$tmp_dir/kubernetes-nmstate
 export ARTIFACTS=${ARTIFACTS-$tmp_dir/artifacts}
 mkdir -p $ARTIFACTS
 
+cp $TMP_PROJECT_PATH/test_logs/e2e/*.log $ARTIFACTS
```
Comment (phoracek) [inline automation/check-patch.setup.sh]: This is setup script, it is called before any tests are executed. You have several options:

1. Dump those logs directly to $ARTIFACTS
2. Introduce teardown script
3. Add explicit `cp` to each lane after tests finish

Comment (tsorya) [reply automation/check-patch.setup.sh]: Got it. Added the relevant line to providers scripts.

File: automation/check-patch.setup.sh [hunk 4]
```diff
@@ -9,6 +9,7 @@ rm -rf $tmp_dir
 mkdir -p $tmp_dir
 
 export TMP_PROJECT_PATH=$tmp_dir/kubernetes-nmstate
+export E2E_LOGS=${TMP_PROJECT_PATH}/test_logs/e2e
```
Comment (qinqon) [inline automation/check-patch.setup.sh]: mkdir -p is missing ?

File: test/e2e/utils.go [hunk 5]
```diff
@@ -40,39 +42,70 @@ var (
...
+			defer podLogs.Close()
+			rawLogs, err := ioutil.ReadAll(podLogs)
+			if err != nil {
+				io.WriteString(GinkgoWriter, fmt.Sprintf("error reading kubernetes-nmstate logs: %v\n", err))
+				continue
+			}
+			formattedLogs := strings.Replace(string(rawLogs), "\\n", "\n", -1)
+			io.WriteString(writer, formattedLogs)
+			// If test failed print pods logs to screen
+			if testFailed{
```
Comment (qinqon) [inline test/e2e/utils.go]: Ginkgo is already not printing GinkgoWriter if test fail, but we disable it by passing "ginkgo.v" maybe we want to remove it again.

File: automation/check-patch.setup.sh (hunk 4 above)
Comment (tsorya) [reply automation/check-patch.setup.sh]: It is part of make file

File: test/e2e/utils.go (hunk 5 above)
Comment (tsorya) [reply test/e2e/utils.go]: Not sure , to say the truth. There parts of the logs that are needed to be printed

File: test/e2e/utils.go [hunk 6]
```diff
@@ -40,39 +42,70 @@ var (
...
-		rawLogs, err := ioutil.ReadAll(podLogs)
-		if err != nil {
-			io.WriteString(writer, fmt.Sprintf("error reading kubernetes-nmstate logs: %v\n", err))
-			continue
-		}
-		formattedLogs := strings.Replace(string(rawLogs), "\\n", "\n", -1)
-		io.WriteString(writer, formattedLogs)
-	}
-	return nil
+	output, err = runAtNode(node, "nmcli", "d", "s")
```
Comment (romfreiman) [inline test/e2e/utils.go]: full path?

File: test/e2e/utils.go [hunk 7]
```diff
@@ -40,39 +42,70 @@ var (
...
+	GinkgoWriter.Write([]byte(fmt.Sprintf("\n***** Done ip status on node %s *****\n", node)))
+}
+
+func writeNetworkManagerLogs(node string, sinceSeconds int){
+	OpenTestLogFile("netmanger", func(writer io.ReadWriteCloser) {
+		output, err := runAtNode(node, "sudo", "journalctl", "-u", "NetworkManager",
+			"--since", fmt.Sprintf("'%ds ago'", sinceSeconds))
+		Expect(err).ToNot(HaveOccurred())
+		writer.Write([]byte(fmt.Sprintf("\n***** Journalctl for net manager on node %s *****\n\n %s", node, output)))
+		writer.Write([]byte(fmt.Sprintf("\n***** Done net manager logs on node %s *****\n", node)))
```
Comment (romfreiman) [inline test/e2e/utils.go]: NetworkManager

File: test/e2e/utils.go [hunk 8]
```diff
@@ -534,7 +567,30 @@ func defaultRouteNextHopInterface(node string) AsyncAssertion {
...
 	}, 15*time.Second, 1*time.Second)
 }
+
 func vlan(node string, iface string) string {
 	vlanFilter := fmt.Sprintf("interfaces.#(name==\"%s\").vlan.id", iface)
 	return gjson.ParseBytes(currentStateJSON(node)).Get(vlanFilter).String()
 }
+
+func OpenTestLogFile(logType string, cb func(f io.ReadWriteCloser)){
+	name := fmt.Sprintf("test_logs/e2e/%s_%s.log", getTestName(), logType)
```
Comment (romfreiman) [inline test/e2e/utils.go]: maybe to add timestamp to the file?

Comment (tsorya) [reply test/e2e/utils.go]: Thought about it, but no need for it.

**Comment by qinqon** [comment]
@tsorya, maybe this is a good PR to start to move things out of utils.go and put reporters under reporter/nm.go, reporter/pod.go, etc...

But this extra mile, it we can do that at following PR.

File: test/e2e/utils.go (hunk 5 above)
Comment (qinqon) [reply test/e2e/utils.go]: Ack

File: test/e2e/utils.go (hunk 6 above)
Comment (dankenigsberg) [reply test/e2e/utils.go]: why is a full path better? Depending on a proper definition of PATH is not bad imho. I was once convinced that full paths are somewhat safer (you are not exposed to someone adding his own ~/bin dir into your PATH) but I don't consider it a real threat here.

Comment (tsorya) [reply test/e2e/utils.go]: will change just for matter of consistency in the same function :)

**Comment by tsorya** [comment]
@qinqon @phoracek  maybe the better way is to create one log file per test and just write all relevant data there?

**Comment by phoracek** [comment]
@tsorya it would be really difficult to navigate that. I would prefer to have one file per test (or per failed test). That is way easier to navigate.

**Comment by tsorya** [comment]
@phoracek thought the same, will change the PR

**Comment by tsorya** [comment]
closing will reopen with new reporter pr

**Pull request closed without merging**